pvlib/modelchain.py

```diff
@@ -408,6 +420,41 @@ def sapm(self):
 
         return self
 
+    def desoto(self):
+        (photocurrent, saturation_current, resistance_series,
+         resistance_shunt, nNsVth) = (
+            self.system.calcparams_desoto(self.effective_irradiance,
+                                          self.temps['temp_cell']))
+
+        self.diode_params = (photocurrent, saturation_current, resistance_series,
+                       resistance_shunt, nNsVth)
+
+        self.dc = self.system.singlediode(
+            photocurrent, saturation_current, resistance_series,
+            resistance_shunt, nNsVth)
+
+        self.dc = self.system.scale_voltage_current_power(self.dc).fillna(0)
+
+        return self
+
+    def pvsyst(self):
+        (photocurrent, saturation_current, resistance_series,
+         resistance_shunt, nNsVth) = (
+            self.system.calcparams_pvsyst(self.effective_irradiance,
+                                          self.temps['temp_cell']))
+
+        self.diode_params = (photocurrent, saturation_current, resistance_series,
+                       resistance_shunt, nNsVth)
+
+        self.dc = self.system.singlediode(
+            photocurrent, saturation_current, resistance_series,
+            resistance_shunt, nNsVth)
+
+        self.dc = self.system.scale_voltage_current_power(self.dc).fillna(0)
+
+        return self
```

Comment:
I don't love the very similar copy/paste code. But it's explicit and that is important too. So I lean towards this approach.

Reply:
I would mark this with `FIXME`. IMO the copy/paste only seems explicit, but it is WET code. DRY code is robust and not prone to errors. This might be maintainable in the short term, but may become a burden in the future, especially if more "models" are added, such as `pvsyst_v7`, ...

There are only two lines of difference between these:
```patch
--- desoto.py
+++ pvsyst.py
@@ -1,7 +1,7 @@
-def desoto(self):
+def pvsyst(self):
     (photocurrent, saturation_current, resistance_series,
      resistance_shunt, nNsVth) = (
-        self.system.calcparams_desoto(self.effective_irradiance,
+        self.system.calcparams_pvsyst(self.effective_irradiance,
                                       self.temps['temp_cell']))
 
     self.diode_params = (photocurrent, saturation_current,

```

You could add a tiny dispatcher that you delegate this common task to that is DRY and also explicit:

```python
def desoto(self):
    """calculate desoto model"""
    return self._model_dispatcher(self.system.calcparams_desoto, 'desoto')

def pvsyst(self):
    """calculate pvsyst model"""
    return self._model_dispatcher(self.system.calcparams_pvsyst, 'pvsyst')

def _model_dispatcher(self, model_calcparams, model_name=None):
    """
    Delegate that dispatches the model

    Parameters
    ----------
    model_calcparams : function
        A function that calculates the models parameters given irradiance and cell temperature.
    model_name : str
        Optional string that may be used to further distinguish the model in the future.
    """
    (photocurrent, saturation_current, resistance_series,
     resistance_shunt, nNsVth) = (
        model_calcparams(self.effective_irradiance,
                                      self.temps['temp_cell']))

    self.diode_params = (photocurrent, saturation_current,
                         resistance_series,
                         resistance_shunt, nNsVth)

    self.dc = self.system.singlediode(
        photocurrent, saturation_current, resistance_series,
        resistance_shunt, nNsVth)

    self.dc = self.system.scale_voltage_current_power(self.dc).fillna(0)

    return self
```

Reply:
Either approach is ok with me at this point. Keep in mind that we've had discussions elsewhere about whether or not it even makes sense to define these model-specific methods on ``ModelChain`` and if we could replace them with wrappers on ``PVSystem``. So I'm not too concerned about perfection here.

Reply:
okay, me too, I fine either way 😁

Reply:
@mikofski that's a good suggestion.  You are right that we have more diode models waiting - CEC hasn't been connected up yet.  But we will (at least I hope) also have `system.doublediode` someday, and perhaps other variations on the calculate parameters -> calculate IV curve pattern.

Reply:
Let's keep the bite small, merge this, and take up improvements in another pull request.